File: MattEland.Emergence.LevelGen/RoomPlacement.fs
module MattEland.Emergence.Domain.RoomPlacement

open MattEland.Emergence.Model
open MattEland.Emergence.LevelData

let mergeChars oldChar newChar =
    match oldChar with
    | '+' -> oldChar
    | _ -> match newChar with
            | ' ' -> oldChar
            | _ -> newChar

type RoomPlacement (room: RoomData, upperLeftCorner: Position) =    
    member this.getCharAtPos (pos: Position) currentChar =
        pos.Subtract upperLeftCorner |> room.GetCharacterAtPosition |> mergeChars currentChar 

File: MattEland.Emergence.LevelData/RoompLacement.cs
using MattEland.Emergence.Model;

namespace MattEland.Emergence.LevelData
{
    public class RoomPlacement
    {
        private readonly RoomData _room;
        private readonly Position _upperLeftCorner;

        public RoomPlacement(RoomData room, Position upperLeftCorner)
        {
            _room = room;
            _upperLeftCorner = upperLeftCorner;
        }

        public char GetChar(Position pos, char currentChar)
        {
            var relativePos = pos.Subtract(_upperLeftCorner);
            char roomChar = _room.GetCharacterAtPosition(relativePos);
            return MergeChars(currentChar, roomChar);
        }

        private static char MergeChars(char oldChar, char newChar) => oldChar == '+' || newChar == ' ' ? oldChar : newChar;
    }

}